src/redirect/server.js

```diff
@@ -0,0 +1,158 @@
+*/
+app.post('/qr', async (req, res) => {
+  logger.info('Sending POST request to /qr');
+  console.log(process.env.PORT);
+  const qrId = randomId();
+  // set mapping of ID to object containing QR details
+  QRs.set(qrId, {
+    id: qrId,
+    request: req.body,
+    claims: []
+  });
+
+  // generate and return url for the authorization server
+  res.json({
+    url: `${PUBLIC_URL}/qr/${qrId}/claim`
+  });
+});
+
+/*
+POST to .url property of the JSON payload to redeem shclink.
+
+If the request is valid, server returns 301 status and Location header redirecting
+client to URL where file/Sharing Manifest is exposed.
+*/
+app.post('/qr/:id/claim', (req, res) => {
+  const qrId = req.params.id;
+  logger.info(`Sending POST request to /qr/${qrId}/claim`);
+  // Get QR details using given QR id
+  const policy = QRs.get(qrId);
+
+  if (!policy || policy.claims.length >= (policy.request.claimsLimit || Infinity)) {
+    res.status(403);
+    return res.json(`QR ${qrId} is not valid or has already been claimed`);
+  }
+
+  const clientSpecificUrl = `/qr/${qrId}/claimed/${randomId()}`;
+
+  // create new claim
+  policy.claims.push({
+    clientName: req.query.clientName || 'unknown',
+    clientSpecificUrl,
+    queryLog: [`Claimed: ${new Date()}`],
+    locationAlias: Object.fromEntries(
+      Array.from(new Set(policy.request.access.flatMap(a => a.locations || [])).values()).map(l => [l, randomId()])
+    )
+  });
+
+  // redirect to 301 Moved Permanently if redemption is successful
+  res.redirect(301, clientSpecificUrl);
+});
+
+/*
+GET to clientSpecificURL for SHCLink that was claimed.
+
+Allows client to access location where SHCLink file/Sharing Manifest is exposed.
+*/
+app.get('/qr/:id/claimed/:cid', (req, res) => {
+  const { cid, id } = req.params;
+  logger.info(`Sending GET request to /qr/${id}/claimed/${cid}`);
+  const policy = QRs.get(id);
+
+  if (!policy) {
+    res.status(403);
+    return res.json(`QR ${id} is no longer valid`);
+  }
+
+  const claimDetails = policy.claims.find(c => c.clientSpecificUrl === req.url);
+  if (!claimDetails) {
+    res.status(403);
+    return res.json(`This QR has not been correctly claimed`);
+  }
+
+  res.json(
+    policy.request.access.map(a => ({
+      ...a,
+      locations: a.locations?.map(l => `${PUBLIC_URL}/qr/${id}/claimed/${cid}/files/${claimDetails.locationAlias[l]}`)
+    }))
+  );
+});
+
+/*
+GET claimed file.
+*/
+app.get('/qr/:id/claimed/:cid/files/:fileid', async (req, res) => {
+  const { id, cid, fileid } = req.params;
+  logger.info(`Sending GET request to /qr/${id}/claimed/${cid}/files/${fileid}`);
+  const policy = QRs.get(id);
+
+  if (!policy) {
+    res.status(403);
+    return res.json(`QR ${id} is no longer valid`);
+  }
+
+  const claimDetails = policy.claims.find(c => c.clientSpecificUrl === `/qr/${id}/claimed/${cid}`);
+  if (!claimDetails) {
+    res.status(403);
+    return res.json(`This QR has not been correctly claimed`);
```

Comment:
There are a few times in this file where we input a string into the `res.json` property. It seems a bit weird sending back a response of JSON content-type with incorrectly formatted JSON. Should we change these instances to `res.send`?

See: https://expressjs.com/en/api.html#res

Reply:
Good point, I think `res.send` and `res.json` are very similar except `res.json` formats the body to JSON and then calls `res.send` under the hood. But since we are just sending back strings I don't think it makes sense to apply the JSON formatting. I'll change these instances to `res.send` instead.